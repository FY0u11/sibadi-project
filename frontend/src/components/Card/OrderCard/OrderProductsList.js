import React from 'react'

const OrderProductsList = ({ orderProducts }) => {
  return (
    <div>
      <h5>Product list</h5>
      <ol>
        {orderProducts.map(product => (
          <li key={product.id}>
            {product.name} <span className="small text-muted">{product.price} $</span>
          </li>
        ))}
      </ol>
    </div>
  )
}

export default OrderProductsList
