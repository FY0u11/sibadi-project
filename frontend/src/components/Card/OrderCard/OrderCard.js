import React from 'react'

import OrderProductsList from './OrderProductsList'
import ExpandingCard from '../ExpandingCard'

const OrderCard = ({
  isCancelButtonShown,
  body,
  orderProducts,
  order,
  isExpanded,
  askForCancellingOrder
}) => {
  const image = <img src={`/uploads/${order.imagePath}`} className="card-img" alt="..." />
  const innerTitle = (
    <div>
      Order #{order.id}
      {isCancelButtonShown && (
        <button className="close text-danger" onClick={askForCancellingOrder}>
          &times;
        </button>
      )}
    </div>
  )
  const innerBody = (
    <div>
      <div>Total price: $ {order.price}</div>
      <div>
        Status:&ensp;
        <strong className={order.status === 'cancelled' ? 'text-danger' : 'text-success'}>
          {order.status}
        </strong>
      </div>
      {body}
    </div>
  )
  const exBody = <OrderProductsList orderProducts={orderProducts} />

  return (
    <ExpandingCard
      image={image}
      title={innerTitle}
      body={innerBody}
      exBody={exBody}
      isExpanded={isExpanded}
    />
  )
}

export default OrderCard
