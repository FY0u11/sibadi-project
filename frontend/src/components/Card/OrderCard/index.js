import React, { useState, useEffect } from 'react'
import { connect } from 'react-redux'

import { updateOrderRequest } from '../../../utils/api'
import { setUser } from '../../../store/actions'
import { updateOrder } from '../../../store/scenes/orders/actions'
import { setModal, closeModal } from '../../../store/components/modal/actions'
import { setNotificationError } from '../../../store/components/notifications/actions'

import Button from '../../Button'
import OrderCard from './OrderCard'

const OrderCardContainer = ({
  user,
  setUser,
  showCancelWhen,
  body,
  isExpanded,
  order,
  updateOrder,
  setModal,
  closeModal,
  products,
  setNotificationError
}) => {
  const [orderProducts, setOrderProducts] = useState([])
  useEffect(() => {
    const orderProductsTmp = []
    order.orderProducts.forEach(orderProductId => {
      const product = products.find(product => product.id === orderProductId)
      orderProductsTmp.push(product)
    })
    setOrderProducts(orderProductsTmp)
  }, [])

  const cancelOrder = async () => {
    try {
      await updateOrderRequest({ status: 'cancelled' }, order.id)
      updateOrder(order.id, { status: 'cancelled' })
      setUser({ ...user, budget: user.budget + +order.price })
      closeModal()
    } catch (error) {
      setNotificationError(error.message)
    }
  }

  const askForCancellingOrder = () => {
    setModal({
      title: 'This cancels the order',
      body: (
        <div style={{ padding: '10px' }}>
          <div style={{ marginBottom: '20px' }}>Are you sure?</div>
          <Button action={cancelOrder}>Yes</Button>
          <Button action={closeModal}>No</Button>
        </div>
      )
    })
  }

  return (
    <OrderCard
      isCancelButtonShown={showCancelWhen.includes(order.status)}
      body={body}
      isExpanded={isExpanded}
      orderProducts={orderProducts}
      order={order}
      askForCancellingOrder={askForCancellingOrder}
    />
  )
}

const putStateToProps = state => ({
  user: state.root.user,
  products: state.products.products
})

const putActionsToProps = { setUser, updateOrder, setModal, closeModal, setNotificationError }

export default connect(putStateToProps, putActionsToProps)(OrderCardContainer)
