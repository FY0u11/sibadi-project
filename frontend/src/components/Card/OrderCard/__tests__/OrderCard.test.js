import React from 'react'
import { render, cleanup } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import OrderCard from '../OrderCard'

afterEach(cleanup)

const props = {
  isCancelButtonShown: true,
  body: 'test body',
  orderProducts: [
    { id: 2, name: 'product 1', price: 10 },
    { id: 3, name: 'product 2', price: 20 },
    { id: 5, name: 'product 3', price: 30 }
  ],
  order: {
    id: 42,
    price: 60,
    status: 'created',
    imagePath: 'path/1'
  },
  isExpanded: true
}

describe('Testing OrderCard component', () => {
  it('It should be properly rendered', () => {
    const { asFragment } = render(<OrderCard {...props} />)
    expect(asFragment()).toMatchSnapshot('1')
  })
  it('It should be properly rendered if order cancelled', () => {
    props.order.status = 'cancelled'
    props.isCancelButtonShown = false
    const { asFragment } = render(<OrderCard {...props} />)
    expect(asFragment()).toMatchSnapshot('2')
  })
})
