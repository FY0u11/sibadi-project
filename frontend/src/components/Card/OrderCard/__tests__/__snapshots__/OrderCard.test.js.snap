// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing OrderCard component It should be properly rendered if order cancelled: 2 1`] = `
<DocumentFragment>
  <div
    class="my-expanding-card"
    style="width: 500px;"
  >
    <div
      class="card my-expanded"
    >
      <div
        class="row no-gutters"
      >
        <div
          class="col-md-4"
        >
          <img
            alt="..."
            class="card-img"
            src="/uploads/path/1"
          />
        </div>
        <div
          class="col-md-8"
        >
          <div
            class="card-body"
          >
            <h5
              class="card-title"
            >
              <div>
                Order #42
              </div>
            </h5>
            <div>
              <div>
                Total price: $ 60
              </div>
              <div>
                Status: 
                <strong
                  class="text-danger"
                >
                  cancelled
                </strong>
              </div>
              test body
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="my-expanded-body"
    >
      <div>
        <h5>
          Product list
        </h5>
        <ol>
          <li>
            product 1 
            <span
              class="small text-muted"
            >
              10 $
            </span>
          </li>
          <li>
            product 2 
            <span
              class="small text-muted"
            >
              20 $
            </span>
          </li>
          <li>
            product 3 
            <span
              class="small text-muted"
            >
              30 $
            </span>
          </li>
        </ol>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Testing OrderCard component It should be properly rendered: 1 1`] = `
<DocumentFragment>
  <div
    class="my-expanding-card"
    style="width: 500px;"
  >
    <div
      class="card my-expanded"
    >
      <div
        class="row no-gutters"
      >
        <div
          class="col-md-4"
        >
          <img
            alt="..."
            class="card-img"
            src="/uploads/path/1"
          />
        </div>
        <div
          class="col-md-8"
        >
          <div
            class="card-body"
          >
            <h5
              class="card-title"
            >
              <div>
                Order #42
                <button
                  class="close text-danger"
                >
                  ×
                </button>
              </div>
            </h5>
            <div>
              <div>
                Total price: $ 60
              </div>
              <div>
                Status: 
                <strong
                  class="text-success"
                >
                  created
                </strong>
              </div>
              test body
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="my-expanded-body"
    >
      <div>
        <h5>
          Product list
        </h5>
        <ol>
          <li>
            product 1 
            <span
              class="small text-muted"
            >
              10 $
            </span>
          </li>
          <li>
            product 2 
            <span
              class="small text-muted"
            >
              20 $
            </span>
          </li>
          <li>
            product 3 
            <span
              class="small text-muted"
            >
              30 $
            </span>
          </li>
        </ol>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
